Reworked High Level Design (#65)
Opster/opensearch-k8s-operator: docs/designs/high-level.md

      end
      subgraph controller [Controller]
          direction TB
          opensearch-reconciler[/OpenSearch Reconciler function/] --Create & execute -->reconcilers
          subgraph reconcilers [Reconcilers Chain]
              reconciler1[Reconciler 1] --- reconciler2
              reconciler2[Reconciler 2] --- reconciler3
              reconciler3[Reconciler 3] --- reconciler_etc[...]
          end
          subgraph components
              scaler([Scaler component])
              auto-scaler([Auto-Scaler Component]) 
          end
      end
      
      controller-- watches -->cr{{OpenSearchCluster Custom Resource}}
      
      subgraph kubernetes-objects-created [Created K8s Objects]
          direction LR
          opensearch-cluster-statefulsets[OpenSearch Cluster Statefulsets]
          openseach-config-map[OpenSearch ConfigMap]
          opensearch-service[OpenSearch Service]
          ...
      end    

      subgraph monitoring [Monitoring Tools]
          prometheus[Prometheus]
          grafana[Grafana]
          alert-manager[Alert Manager]
      end
      
      manager --> kubernetes-objects-created
```

### Provisioning 

> The operator contains a Manager, which manges a Controller for its Custom Resource (`OpenSearchCluster`). It is mainly a listener on queue, which gets filled when ever a change has been made to one of the custom resources. The action is adjusting the deployed cluster to match the desired configuration specified in the custom resource. It can be a simple change to the Security configuration which may only update dynamic settings using OpenSearch API,or it can also be a change to the pod spec, increasing memory requirements from 10gb to 12gb. The latter requires a series of steps to accomplish that.

Reviewer: Typo: manges -> manages

> Once a change has been detected it activates the Controller with the changed custom resource ID, which activates the OpenSearch Reconciler reconcile function. That reconciler is the main entry for the whole operator. Its goal is to modify/create/delete the OpenSearch Cluster and other deployed tools (e.g. Prometheus) reflecting the changes made to the custom resource. 

Reviewer: I prefer "call" instead of "activate". "activate" for me sounds more like some indirect/async interaction.

The OpenSearch Reconciler reconcile function is primarily composed of two types of elements: Reconcilers and Components. The Reconcilers main job is to update/create the deployed Kubernetes objects representing the OpenSearch Cluster and other various software the Operator needs. If we take the Custom Resource YAML and split into it logical parts (i.e. fields), then a Reconciler is mapped to a specific part (field). The operators have the following Reconcilers:
* Cluster Reconciler: mapped to `nodePools`. In charge of OpenSearch cluster
* Dashboards Reconciler: mapped to `dashboards`. In charge of [OpenSearch Dashboards](https://opensearch.org/docs/latest/dashboards/index/) 
* Configuration Reconciler - In charge of OpenSearch configuration
* TLS Reconciler - mapped to `security.tls`. In charge of TLS configuration (OpenSearch Security Plugin)

> 

Reviewer: Missing here: Securityconfig Reconciler - mapped to `security.config`. In charge of the securityconfig with users and roles (OpenSearch Security plugin)

The Reconcilers and Components communicate via a `ReconcilerContext` that is passed between them which they can use to share information during one reconcile run (for example the TLS reconciler uses it to tell the cluster reconciler to mount secrets containing TLS certificates to the pods). To communicate between runs and persist state the Status object of the Custom Resource can be used. It contains a data structure that exposes the current status of each reconciler and component. This has the double advantage of sharing state between Reconcilers and also make it visible to the user.

> The OpenSearch Reconciler retrieves the current custom resource, based on the ID, from Kubernetes (using its API) and checks to see whether it was deleted (by checking if the custom resource has deletion timestamp field in it). If it was, it removes the Finalizer that was placed on it, and calls `DeleteResources` for each Reconciler. All Kubernetes objects created by the operator are using [Owner Reference](https://kubernetes.io/blog/2021/05/14/using-finalizers-to-control-deletion/#owner-references) to the custom resource, thus are deleted in a cascading fashion by Kubernetes once the finalizer has been removed. 

Reviewer: The order is different: First the `DeleteResources` methods are called, then the finalizer is removed as a last step. And maybe add a short explanation that in that method the Reconcilers deal with any cleanup that cannot be handled implicitly via the Owner References

If it was created or updated, the OpenSearch Reconciler uses the retrieved custom resource YAML to parse it and produces a data structure representing it. It then executes the chain of Reconcilers, passing the Custom Resource data structure as main input.

Each Reconciler executes and returns a special Result structure, that says two things:
* Requeue: true/false. If true, it will fill the queue the controller is listening on with a message with the same resource ID. False - it won't do it, hence the OpenSearch Reconciler will only get called upon another change to the custom resource

> * RequeueAfter: units is seconds. If Requeue is true, the message will be inserted into the queue in RequeueAfter seconds.

Reviewer: Technically the type is `time.Duration` so we can use whatever unit we want (e.g. `30 * time.Second`)

> If one of the Reconcilers returns an error, or states Requeue (true), then the OpenSearch Reconciler returns the same result object, meaning the OpenSearch Reconciler will be called again in RequeueAfter seconds, calling the chain of reconcilers again.

Reviewer: ```suggestion
If one of the Reconcilers returns an error, or requests a Requeue (via the Result object), then the OpenSearch Reconciler returns the same result object, meaning the OpenSearch Reconciler will be called again in RequeueAfter seconds, calling the chain of reconcilers again.
```

Regardless of the Requeue (true/false) returned, the reconciler updates its status in the Status element of the custom resource.

> Notice that Reconcilers in the reconcilers chain, do not save any state in them. They get instantiated on each call to the OpenSearch Reconciler reconcile function.

Reviewer: ```suggestion
Notice that Reconcilers in the reconcilers chain do not save any state in their struct instance. They get instantiated on each call to the OpenSearch Reconciler reconcile function so must be treated as stateless and must store any state to be persisted between runs in the Status object.
```
